### ISSUE TITLE
no error for invalid protocol match with `Self` parameter

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#2818]
### Summary

```py
from typing import Self, Protocol

class CanFoo(Protocol):
    def foo(self, other: Self):
        pass


class CanFoo1(CanFoo):
    def foo(self, other: Self):  # error: invalid override
        pass


class CanFoo2:
    i = 1
    def foo(self, other: Self):
        self.i + other.i

_: CanFoo = CanFoo2()  # no error


def f[T: CanFoo](t1: T, t2: T):
    t1.foo(t2)  # runtime error

f(CanFoo1(), CanFoo2())  # no error
```

**Comment by carljm** [comment]
Yes, thanks -- a method of a concrete type accepting `Self` is not a valid implementation of a protocol method accepting `Self`. (I think to be valid, the `other` argument of `CanFoo2.foo` would need to be typed as `CanFoo`.

**Comment by charliermarsh** [comment]
Am I misunderstanding this test in the conformance suite? https://github.com/python/typing/blob/ffd520acd116aacbbba793c8f1f1fd51dfda83ae/conformance/tests/protocols_self.py#L72

**Comment by KotlinIsland** [comment]
> Am I misunderstanding this test in the conformance suite? https://github.com/python/typing/blob/ffd520acd116aacbbba793c8f1f1fd51dfda83ae/conformance/tests/protocols_self.py#L72

**edit**: i fixed the example

looks invalid to me, it's inconsistent across classes and protocols. the autor of that conforemance test also implemented this feature in pyright:
Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYA0UAygKYXA1gBu9IISAJvQFB9SFAIYBnUVACCALj5R5UXsCjAADAApRjZlAD6XNAAsY0utoCUAWgB8UAHJgU9UwDo3AoWIkAhdZPOyClAAAmwcXLxyCkoqGlpMNPpIRiZmTFa2Dk6ubvIAxFDhuKYARJIlmBIoYPheySjCAEYU9LBgUCXeJQKCIuKEAEwEwiD0KDDqRDVklAFR8jFqmtqJBsamDOk29o7OUG4uPZ79AMIDgdH0ykvxOkkpGxZQ21l7B-lQ1YWcuD3C54NhqNxlAALxQM7qcwfL5FEAePoSE4AZkmQxGYxgcyCoXYnB4-CCiziKz0a1Sm2A0JeuxyLhh7ThAiAA)

```python
from typing import Protocol, Self, override

class A:
    def f0(self, _right: Self)-> None: ...

class B(A):
    @override
    def f0(self, _right: Self)-> None: ...  # error: "A" is not assignable to "B"


class P2Parent(Protocol):
    def f0(self, _right: Self)-> None: ...


class C2:
    def f0(self, _right: Self) -> None: ...


a2: P2Parent = C2()  # no error

class C3(P2Parent):
    @override
    def f0(self, _right: Self) -> None: ...  # no error


```

**Comment by carljm** [comment]
> Am I misunderstanding this test in the conformance suite?

I think the conformance suite is asserting something unsound there. As shown in the OP example here, if that is allowed then two different types could both implement `P2Parent`, and be passed to a function taking in `(x: P2Parent, y: P2Parent)` and the type-checker would allow the call `x.f0(y)` -- but if `x` and `y` are different concrete types, that call is invalid (`y` should not satisfy `Self` on `x`).

This is really just a variant of the general fact that `Self` typed arguments other than `self` aren't safe for inheritance.

I think before we make this change we should suggest a conformance suite change that would allow type checkers to error here.

@KotlinIsland I'm not sure what the basedpyright playground link is supposed to demonstrate. It looks like basedpyright is even more forgiving/unsound than ty here, allowing both the inheritance (other than wanting `@override` marked) and the assignment?

**Comment by KotlinIsland** [comment]
> @KotlinIsland I'm not sure what the basedpyright playground link is supposed to demonstrate. It looks like basedpyright is even more forgiving/unsound than ty here, allowing both the inheritance (other than wanting @override marked) and the assignment?

i was demonstrating what you are describing, that it is unsafe and inconsistent and broken in pyright as well

the maintainer of pyright was the one that implemented this logic and also wrote the conformance test

edit: oh, i noticed that my playground was completely nonsensical, i too quickly read the error message and assumed it was about the `Self` parameter

**Comment by KotlinIsland** [comment]
> other than `self` aren't safe for inheritance

it's not generally safe for `self` either:

```py
from typing import override


class A:
    def f(self) -> None:
        pass


class B(A):
    a = 1
    
    @override
    def f(self) -> None:
        self.a


def foo(t: type[A]):
    t.f(A())


foo(B)
```

although maybe this is a comment on the unsafety of `type`, not `self`

**Comment by carljm** [comment]
> it's not generally safe for `self` either

Yes, I'm aware. Though since `Self` typing of `self` is kind of unavoidable, I prefer to frame this as "calling unbound methods on `type[]` types is unsound"
